datafusion/datasource/src/source.rs

```diff
@@ -267,7 +269,23 @@ impl ExecutionPlan for DataSourceExec {
         partition: usize,
         context: Arc<TaskContext>,
     ) -> Result<SendableRecordBatchStream> {
-        self.data_source.open(partition, Arc::clone(&context))
+        let stream = self.data_source.open(partition, Arc::clone(&context))?;
+        let batch_size = context.session_config().batch_size();
+        let threshold = context.session_config().batch_split_threshold();
+        if threshold > 0 && batch_size >= threshold {
```

Comment:
As I mentioned above, I think it would be better if we always used the `BatchSplitStream` (or inlined the logic into MemorySource)